### PR TITLE
add 1 homework

### PR DESCRIPTION
**Pull request by krastykovyaz**
В работе реализован однослойный линейный классификатор - перцептрон. Для начала необходимо установить библиотеки из папки requirements.txt><br />
 Реализованы следующие флаги при запуске: ```+1``` <br />
 usage: ft_perceptron.py [-h] [-t] [-p] [--plot] [--valid-size VALID_SIZE] [-lr LEARNING_RATE] [-e EPOCHS]<br />
                         [--batch-size BATCH_SIZE] [--train-path TRAIN_PATH] [--test-path TEST_PATH]<br />
 optional arguments:<br />
```
  -h, --help            show this help message and exit
  -t, --train           train mode
  -p, --predict         predict mode
  --plot                visualization
  --valid-size VALID_SIZE
                        split validation size (0-1)
  -lr LEARNING_RATE, --learning-rate LEARNING_RATE
  -e EPOCHS, --epochs EPOCHS
                        number epochs
  --batch-size BATCH_SIZE
  --train-path TRAIN_PATH, --train-filename TRAIN_PATH
                        train dataset filename
  --test-path TEST_PATH, --test-filename TEST_PATH
                        test dataset filename
```
Проект полностью логируется в режимах обучения и тестирования.<br />
Также реализован режим отображения изменения loss во время обучения. ```+1```<br />
Также во время обучения можно остановить процесс, что позволит сохранить модель с текущими весами. То есть это остановит процесс обучения, но при этом не сломает саму модель. ```+1```<br />
Можно передать в аргументах количество эпох для обучения, также задать learning_rate, и  batch_size, а также размер датасета для вадидации valid_size. ```+1```<br />

0. Создан pull request c полным описанием проделанной работы. ```+1```<br />
1. Сделан отчет по каждому пункту из задания. ```+1```<br />
2. Проведен EDA анализ. Ноутбук лежит в папке ноутбукс. Также реализована программа по формированию отчета в pdf формате. ```+1```<br />
Запускать под средой в папке scripts - python3 gen_eda_report.py или ./eda.sh <br />
Отчет сохранится в папку pics ```+1```<br />
3. Написана функция/класс для тренировки модели, вызов оформлен как утилита командной строки, записана в readme инструкцию по запуску
Создана функция для тренировки модели train. Вызов осуществляется следующей командой:<br />
python3 ft_perceptron.py или в корне ./train.sh ```+3```<br />
4. Написана функция/класс predict (вызов оформлен как утилита командной строки), которая примет на вход артефакт/ы от обучения, тестовую выборку (без меток) и запишет предикт по заданному пути, инструкция по вызову записана в readme <br />
Создана функция для тестирования модели predict. Вызов осуществляется следующей командой:<br />
python3 ft_perceptron.py --predict или в корне ./test.sh ```+3```<br />
5. Проект имеет модульную структуру ```+2```<br />
6. Использованы логгеры. Все процессы логируются в файл file.log ```+2```<br />
7. ~~Написаны тесты на отдельные модули и на прогон обучения и predict (3 балла)~~
8. ~~Для тестов генерируются синтетические данные, приближенные к реальным (2 балла)~~
9. Обучение модели конфигурируется с помощью конфигов в json или yaml, закоммитьте как минимум 2 корректные конфигурации, с помощью которых можно обучить модель (разные модели, стратегии split, preprocessing)<br />
Json файл с весами и pkl модедью загружается в папку data в файл в папке data model.json. ```+3```<br />
10. ~~Используются датаклассы для сущностей из конфига, а не голые dict (2 балла)~~<br />
11. ~~Напишите кастомный трансформер и протестируйте его (3 балла) https://towardsdatascience.com/pipelines-custom-transformers-in-scikit-learn-the-step-by-step-guide-with-python-code-4a7d9b068156~~<br />
12. ~~В проекте зафиксированы все зависимости (1 балл)~~<br />
13. ~~Настроен CI для прогона тестов, линтера на основе github actions (3 балла). Пример с пары: https://github.com/demo-ml-cicd/ml-python-package~~<br />
14. Залогируйте метрики<br />
Метрики логируюся в файл file.log ```+1```<br />

Итого с учетом дополнительных фитч, которые прикручены в проекте, я насчитал ```21 балл```<br />

### CODE REVIEW

File: app/ml_project/ft_perceptron.py [hunk 1]
```diff
@@ -55,7 +58,7 @@ def plotting(train_losses, val_losses, is_end=False, best_loss=None, best_epoch=
     if is_end:
         plt.show()
 
-def train():
+def train(text):
```
Comment (SVizor42) [inline app/ml_project/ft_perceptron.py:61]: Зачем передавать этот аргумент в train и predict?

File: app/ml_project/ft_perceptron.py [hunk 2]
```diff
@@ -89,8 +92,9 @@ def train():
         plotting(train_log, val_log, is_end=True, best_epoch=nn.best_epoch, best_loss=nn.best_loss)
     nn.use_best_weights()
     nn.export_nn(conf.MODEL_DATA_PATH)
+    return "Model is done !"
 
-def predict():
+def predict(text):
```
Comment (SVizor42) [inline app/ml_project/ft_perceptron.py:97]: К чему этот аргумент?
И не хвататает типа возвращаемого значения

File: app/ml_project/ft_perceptron.py [hunk 3]
```diff
@@ -1,4 +1,7 @@
 import pandas as pd
+import sys
+from pathlib import Path
+sys.path.append(f'{Path(__file__).resolve(strict=True).parent}')
 from source.net import NN, Linear, Relu
```
Comment (SVizor42) [inline app/ml_project/ft_perceptron.py:5]: Локальные импорты следуют после станлартных пакетов

File: app/model/model.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import pandas as pd
+from loguru import logger
+import sys, re
```
Comment (SVizor42) [inline app/model/model.py:3]: Каждый импорт, как правило, должен быть на отдельной строке

**Review by SVizor42 — APPROVED** [review]
**Важно:** PR был переоткрыт взамен старого, оригинальный был открыт в срок и находится тут https://github.com/made-mlops-2022/aleks_kovyazin/pull/1 

Общий комментарий - складывается ощущение, что в эту ветку попали файлы для второго ДЗ, что затрудняет проверку. Нужно стараться избегать в будущем подобных ситуаций.

Кроме того, все последующие изменения/исправления по текущему ДЗ необходимо вносить в существующую ветку. Создавать новую ветку или открывать новый PR не нужно (без необходимости).

По самой работе:
1. Нужно добавить метки `MADE`/`TECHNOPARK` и `hw1`.
2. Параметры слоев нейросетки лучше тоже вынести в конфиг или константы.
3. В продакшн коде лучше избегать использования ассертов в основной логике.
4. Пункт про конфиги в формате json или yaml - речь идет об использовании конфигов для обучения, содержащих пути к данным, стратегии разбиения данных, параметры моделей и т.д. Я обнаружил только один конфиг в формат `py`. За этот пункт 1.5 балла.
5. По поводу логирования метрик - тут речь шла об использовании mlflow. Этот пункт не учитывается.
6. Самооценку нужно проводить на основе существующих пунктов. Решение о каких-либо дополнительных баллах (сверх описанных в ДЗ пунктов) может принимать только ревьювер. Могу добавить 1 балл за визуализацию лосса в ходе обучения и еще 1 балл за возможность сохранения весов модели в случае прерывания обучения.

**Сумма: 15.5 + 2 (доп баллы) = 17.5 баллов**

Ветку можно сливать, если нет возражений.

**Comment by krastykovyaz** [comment]
Спасибо, согласен.

**Comment by SVizor42** [comment]
Нужно было мерджить ветку в мейн, а не закрывать

**Comment by krastykovyaz** [comment]
я смержил в master

**Comment by krastykovyaz** [comment]
закрыл после оценки по просьбе тютора
